app/code/Magento/Inventory/Model/SourceItem/Validator/StatusConsistencyValidator.php

```diff
@@ -0,0 +1,159 @@
+use Magento\CatalogInventory\Api\StockConfigurationInterface;
+use Magento\CatalogInventory\Api\StockItemCriteriaInterfaceFactory;
+use Magento\CatalogInventory\Model\Stock\Item as LegacyStockItem;
+use Magento\CatalogInventory\Model\Stock\StockItemRepository as LegacyStockItemRepository;
+use Magento\Framework\Validation\ValidationResult;
+use Magento\Framework\Validation\ValidationResultFactory;
+use Magento\Inventory\Model\OptionSource\SourceItemStatus;
+use Magento\Inventory\Model\SourceItem\Validator\StatusConsistencyValidator\IsValidationAllowed;
+use Magento\InventoryApi\Api\Data\SourceItemInterface;
+
+/**
+ * Check that if product quantity <= 0 status is not "In Stock".
+ */
+class StatusConsistencyValidator implements SourceItemValidatorInterface
+{
+    /**
+     * @var ValidationResultFactory
+     */
+    private $validationResultFactory;
+
+    /**
+     * @var SourceItemStatus
+     */
+    private $sourceItemStatus;
+
+    /**
+     * @var StockConfigurationInterface
+     */
+    private $stockConfiguration;
+
+    /**
+     * @var LegacyStockItemRepository
+     */
+    private $legacyStockItemRepository;
+
+    /**
+     * @var ProductResourceModel
+     */
+    private $productResource;
+
+    /**
+     * @var StockItemCriteriaInterfaceFactory
+     */
+    private $stockItemCriteriaFactory;
+
+    /**
+     * @var IsValidationAllowed
+     */
+    private $isValidationAllowed;
+
+    /**
+     * @param ValidationResultFactory $validationResultFactory
+     * @param SourceItemStatus $sourceItemStatus
+     * @param StockConfigurationInterface $stockConfiguration
+     * @param LegacyStockItemRepository $legacyStockItemRepository
+     * @param ProductResourceModel $productResource
+     * @param StockItemCriteriaInterfaceFactory $stockItemCriteriaFactory
+     * @param IsValidationAllowed $isValidationAllowed
+     */
+    public function __construct(
+        ValidationResultFactory $validationResultFactory,
+        SourceItemStatus $sourceItemStatus,
+        StockConfigurationInterface $stockConfiguration,
+        LegacyStockItemRepository $legacyStockItemRepository,
+        ProductResourceModel $productResource,
+        StockItemCriteriaInterfaceFactory $stockItemCriteriaFactory,
+        IsValidationAllowed $isValidationAllowed
+    ) {
+        $this->validationResultFactory = $validationResultFactory;
+        $this->sourceItemStatus = $sourceItemStatus;
+        $this->stockConfiguration = $stockConfiguration;
+        $this->legacyStockItemRepository = $legacyStockItemRepository;
+        $this->productResource = $productResource;
+        $this->stockItemCriteriaFactory = $stockItemCriteriaFactory;
+        $this->isValidationAllowed = $isValidationAllowed;
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function validate(SourceItemInterface $sourceItem): ValidationResult
+    {
+        $errors = [];
+        if ($this->isValidationAllowed->execute($sourceItem)) {
```

Comment:
Why do we try to save Source Items for complex products?
Need to investigate and fix it on higher level